### ISSUE TITLE
Supporting font fallbacks for unicode

### ISSUE THREAD

**Issue by StephenLynx** [SFML/SFML#1536]
Currently, SFML supports only a single font for a sf::Text object. So what happens when your string have  multiple unicode characters is that they probably are not going to be rendered, specially if you are talking about non-textual characters, such as ☆.

Now, what can be done is to use a library such as fontconfig to use system fallbacks to render any character that isn't present on the chosen font.

As it stands, sfml text rendering is very limited because of how you can't render arbitrary unicode characters. Imagine a text client, a terminal emulator or a file browser that can't make use of these fallbacks to system fonts to render unicode.

**Comment by StephenLynx** [comment]
Update: as discussed on irc, an alternative would be to have SFML to provide just a means for a programmer to provide a fallback.

Something like passing a function pointer that is used if the character isn't found on the font.

**Comment by MarioLiebisch** [comment]
I actually modified `sf::Font` in the past for personal use to allow the user to add a `sf::Image` as an extra glyph for a code point of the author's choice. In my case I added icons for things such as gamepad button images and items as extra glyphs outside the code range my regular texts used. You're talking about something like that?

**Comment by StephenLynx** [comment]
Yes. For example, if the programmer knows a certain glyph wasn't found on his font, he can then search for the glyph on some other font installed on his system, which is what pango by using fontconfig.

Also, I found out about FT_Get_Char_Index. This function returns 0 if the given character can't be found on the loaded font and seems to be the best way to tell if a fallback is needed to load a character.

**Comment by Edgaru089** [comment]
In my opinion, SFML does not support loading system fonts in the first place, so, instead of loading system fallbacks silently, we should somehow tell the programmer that the glyph is missing in this certain font and let him/her to resolve the issue.

The quick fix to me should be something like a boolean in sf::Glyph to tell if the glyph is really provided in the font (or just a dummy). I might attempt implementing that if it sounds nice.

IMHO, something like a font manager which attempts loading a certain codepoint from a list of fonts one by one sounds nice too.

**Comment by eXpl0it3r** [comment]
With #1644 merged, you can now query the font for valid glyphs. It's not exactly a fallback feature, but you can now achieve something similar, checking some special glyphs and if they don't exist try a different font.

Should we keep this issue open or do we consider this solved, @binary1248 / @LaurentGomila?

**Comment by binary1248** [comment]
What libraries like ImGui do is allow you to check/load entire Unicode ranges beforehand. If you determine, with the help of this new function, that some glyphs would be missing, you can fall back to a system provided Unicode font.

In order to provide ad-hoc switching to a fallback font we would either have to implement some way to merge multiple fonts depending on specific rules or expose a bit more of the `sf::Font` internals. Both don't seem like manageable solutions.
